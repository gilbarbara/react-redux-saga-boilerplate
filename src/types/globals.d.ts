declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__: any;
    HIDE_LOGS: boolean;
    store: any;
  }

  const VERSION: string;
}

export {};
