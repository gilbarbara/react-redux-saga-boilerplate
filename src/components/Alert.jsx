import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import theme from 'modules/theme';

import { Alert as AlertComponent, Box, utils } from 'styled-minimal';
import Icon from 'components/Icon';

const { colors, palette } = utils.themeGet(theme);
const variants = { ...colors, ...palette };

AlertComponent.displayName = 'AlertComponent';

const StyledAlert = styled(AlertComponent)`
  display: flex;
  line-height: 1;
  padding: 0;
  position: relative;
`;

const AlertIcon = styled(Icon)`
  align-items: flex-start;
  background-color: ${({ variant }) => variants[variant]};
  display: flex;
  padding: ${utils.spacer(3)};
`;

const AlertButton = styled.button`
  background-color: ${({ variant }) => variants[variant]};
  pointer-events: all;
  position: absolute;
  right: ${utils.spacer(1)};
  top: ${utils.spacer(1)};
`;

const Alert = ({ children, handleClickClose, id, icon, ...rest }) => {
  const output = {};
  let name;

  switch (rest.variant) {
    case 'success': {
      name = icon || 'check-circle';
      break;
    }
    case 'warning': {
      name = icon || 'exclamation-circle';
      break;
    }
    case 'danger': {
      name = icon || 'times-circle';
      break;
    }
    case 'info': {
      name = icon || 'question-circle';
      break;
    }
    case 'dark': {
      name = icon || 'bell-o';
      break;
    }
    default: {
      name = icon || 'dot-circle-o';
    }
  }

  if (handleClickClose) {
    output.button = (
      <AlertButton data-id={id} onClick={handleClickClose} type="button">
        <Icon name="times" color="#ccc" width={10} />
      </AlertButton>
    );
  }

  return (
    <StyledAlert {...rest}>
      <AlertIcon {...rest} name={name} color="#fff" width={24} />
      <Box p={2} pr={handleClickClose ? 3 : 2}>
        {children}
      </Box>
      {output.button}
    </StyledAlert>
  );
};

Alert.propTypes = {
  children: PropTypes.node.isRequired,
  handleClickClose: PropTypes.func,
  icon: PropTypes.string,
  id: PropTypes.string,
  outline: PropTypes.bool,
  variant: PropTypes.string,
};

Alert.defaultProps = {
  outline: true,
  variant: 'info',
};

export default Alert;
