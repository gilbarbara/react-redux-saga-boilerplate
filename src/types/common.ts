import React from 'react';

export interface ActionCreator<P extends any[] = any[], A = any> {
  (...args: P): StoreAction<A>;
}

export interface ActionsMapReducer<State> {
  [type: string]: (draft: State, action: StoreAction) => any;
}

export type AlertPosition = 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right';

export interface AsyncFlow {
  message: string;
  status: Status;
}

export type GenericFunction<T = any> = (...args: any[]) => T;

export type Icons =
  | 'bell-o'
  | 'bell'
  | 'bolt'
  | 'check-circle-o'
  | 'check-circle'
  | 'check'
  | 'dot-circle-o'
  | 'exclamation-circle'
  | 'question-circle-o'
  | 'question-circle'
  | 'sign-in'
  | 'sign-out'
  | 'times-circle-o'
  | 'times-circle'
  | 'times';

export interface LoggerOptions {
  collapsed?: boolean;
  hideTimestamp?: boolean;
  skip?: boolean;
  typeColor?: string;
}

export type PlainObject<T = any> = Record<string, T>;

export interface RouteProps {
  component: React.ComponentType<any>;
  exact?: boolean;
  isAuthenticated: boolean;
  path: string;
  to?: string;
}

export interface SortFunction {
  (left: PlainObject, right: PlainObject): number;
  (left: string, right: string): number;
}

export interface Status {
  ERROR: 'error';
  IDLE: 'idle';
  READY: 'ready';
  RUNNING: 'running';
  SUCCESS: 'success';
}

export interface StoreAction<T = any> {
  type: string;
  payload?: T;
  meta?: PlainObject;
  error?: boolean;
}

export interface ShowAlertOptions {
  icon?: Icons;
  id?: string;
  position?: AlertPosition;
  timeout?: number;
  variant?: string;
}

export type Transitions = 'fade' | 'slideDown' | 'slideLeft' | 'slideRight' | 'slideUp';
