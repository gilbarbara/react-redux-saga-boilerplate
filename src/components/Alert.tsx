import { MouseEvent, ReactNode } from 'react';
import styled from '@emotion/styled';
import { Alert as AlertComponent, Box, Icon, Types } from '@gilbarbara/components';

import theme from '~/modules/theme';

import { AlertType } from '~/types';

interface Props {
  children: ReactNode;
  handleClickClose?: (event: MouseEvent<HTMLButtonElement>) => void;
  icon?: Types.Icons;
  id?: string;
  type?: AlertType;
}

AlertComponent.displayName = 'AlertComponent';

const AlertButton = styled.button`
  color: #fff;
  line-height: 0;
  pointer-events: all;
  position: absolute;
  right: ${theme.spacing.xs};
  top: ${theme.spacing.xs};
`;

function Alert({ children, handleClickClose, id, type = 'neutral', ...rest }: Props) {
  const output: Record<string, any> = {};

  if (handleClickClose) {
    output.button = (
      <AlertButton data-id={id} onClick={handleClickClose} type="button">
        <Icon name="times" size={10} />
      </AlertButton>
    );
  }

  return (
    <AlertComponent {...rest} align="center" data-component-name="Alert" type={type}>
      <Box pl={handleClickClose ? 'sm' : 'xs'}>{children}</Box>
      {output.button}
    </AlertComponent>
  );
}

export default Alert;
