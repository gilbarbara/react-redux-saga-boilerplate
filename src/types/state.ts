import { Dispatch } from 'redux';
import { Variants } from 'styled-minimal/lib/types';
import { ValueOf } from 'type-fest';

import { AlertPosition, Icons, Status } from './common';

export interface AlertData {
  icon: Icons;
  id: string;
  message: string;
  position: AlertPosition;
  timeout: number;
  variant: Variants;
}

export interface Topic {
  cached: boolean;
  data: Array<Record<string, any>>;
  message: string;
  status: ValueOf<Status>;
  updatedAt: number;
}

export interface AppState {
  alerts: AlertData[];
}

export interface GitHubState {
  topics: Record<string, Topic>;
  query: string;
}

export interface UserState {
  isAuthenticated: boolean;
  status: ValueOf<Status>;
}

export interface StoreState {
  app: AppState;
  github: GitHubState;
  user: UserState;
}

export interface WithDispatch {
  dispatch: Dispatch;
}
