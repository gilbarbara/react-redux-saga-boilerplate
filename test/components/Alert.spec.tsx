import React, { ReactNode } from 'react';
import { render, screen } from 'test-utils';

import Alert from '~/components/Alert';

import { AlertType } from '~/types';

describe('Alert', () => {
  it('should render the default alert', () => {
    render(<Alert>Basic Alert</Alert>);

    expect(screen.getByRole('alert')).toMatchSnapshot();
  });

  it.each<{ children: ReactNode; type: AlertType }>([
    { type: 'success', children: 'This is a success message' },
    { type: 'error', children: <p key="danger">This is an error message</p> },
    { type: 'warning', children: 'This is a warning message' },
    { type: 'info', children: 'This is an info message' },
    { type: 'neutral', children: <div key="dark">This is a message</div> },
  ])('should render the $type type', ({ children, type }) => {
    render(<Alert type={type}>{children}</Alert>);

    expect(screen.getByRole('alert')).toMatchSnapshot();
  });
});
