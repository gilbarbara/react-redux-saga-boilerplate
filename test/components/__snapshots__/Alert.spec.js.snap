// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should render a error alert with markup 1`] = `
.c1 {
  background-color: #fff;
  border: 1px solid #ff4444;
  color: #ff4444;
  border-radius: 3px;
  font-size: 16px;
  line-height: 1.4;
  max-width: 450px;
  padding: 10px 16px;
  width: 100%;
}

.c1 a {
  color: #ff4444;
}

.c4 {
  width: 100%;
  padding: 8px;
  padding-right: 8px;
}

.c3 {
  display: inline-block;
  line-height: 0;
}

.c3 svg {
  height: auto;
  max-height: 100%;
  width: 24px;
}

.c3 svg path {
  fill: #fff;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1;
  padding: 0;
  position: relative;
}

.c2 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  background-color: #ff4444;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 16px;
}

<StyledAlert
  outline={true}
  variant="danger"
>
  <StyledAlert
    className="c0"
    outline={true}
    size="md"
    variant="danger"
  >
    <div
      className="c0 c1"
      role="alert"
      size="md"
    >
      <AlertIcon
        color="#fff"
        name="times-circle"
        outline={true}
        variant="danger"
        width={24}
      >
        <StyledIcon
          className="c2"
          color="#fff"
          outline={true}
          src="/media/icons/times-circle.svg"
          variant="danger"
          width={24}
        >
          <span
            className="isvg loading c2 c3"
          />
        </StyledIcon>
      </AlertIcon>
      <StyledBox
        p={2}
        pr={2}
      >
        <div
          className="c4"
        >
          <p>
            This is an error message
          </p>
        </div>
      </StyledBox>
    </div>
  </StyledAlert>
</StyledAlert>
`;

exports[`Alert should render a info alert 1`] = `
.c1 {
  background-color: #fff;
  border: 1px solid #33b5e5;
  color: #33b5e5;
  border-radius: 3px;
  font-size: 16px;
  line-height: 1.4;
  max-width: 450px;
  padding: 10px 16px;
  width: 100%;
}

.c1 a {
  color: #33b5e5;
}

.c4 {
  width: 100%;
  padding: 8px;
  padding-right: 8px;
}

.c3 {
  display: inline-block;
  line-height: 0;
}

.c3 svg {
  height: auto;
  max-height: 100%;
  width: 24px;
}

.c3 svg path {
  fill: #fff;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1;
  padding: 0;
  position: relative;
}

.c2 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  background-color: #33b5e5;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 16px;
}

<StyledAlert
  outline={true}
  variant="info"
>
  <StyledAlert
    className="c0"
    outline={true}
    size="md"
    variant="info"
  >
    <div
      className="c0 c1"
      role="alert"
      size="md"
    >
      <AlertIcon
        color="#fff"
        name="question-circle"
        outline={true}
        variant="info"
        width={24}
      >
        <StyledIcon
          className="c2"
          color="#fff"
          outline={true}
          src="/media/icons/question-circle.svg"
          variant="info"
          width={24}
        >
          <span
            className="isvg loading c2 c3"
          />
        </StyledIcon>
      </AlertIcon>
      <StyledBox
        p={2}
        pr={2}
      >
        <div
          className="c4"
        >
          This is an info message
        </div>
      </StyledBox>
    </div>
  </StyledAlert>
</StyledAlert>
`;

exports[`Alert should render a success alert 1`] = `
.c1 {
  background-color: #fff;
  border: 1px solid #00C851;
  color: #00C851;
  border-radius: 3px;
  font-size: 16px;
  line-height: 1.4;
  max-width: 450px;
  padding: 10px 16px;
  width: 100%;
}

.c1 a {
  color: #00C851;
}

.c4 {
  width: 100%;
  padding: 8px;
  padding-right: 8px;
}

.c3 {
  display: inline-block;
  line-height: 0;
}

.c3 svg {
  height: auto;
  max-height: 100%;
  width: 24px;
}

.c3 svg path {
  fill: #fff;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1;
  padding: 0;
  position: relative;
}

.c2 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  background-color: #00C851;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 16px;
}

<StyledAlert
  outline={true}
  variant="success"
>
  <StyledAlert
    className="c0"
    outline={true}
    size="md"
    variant="success"
  >
    <div
      className="c0 c1"
      role="alert"
      size="md"
    >
      <AlertIcon
        color="#fff"
        name="check-circle"
        outline={true}
        variant="success"
        width={24}
      >
        <StyledIcon
          className="c2"
          color="#fff"
          outline={true}
          src="/media/icons/check-circle.svg"
          variant="success"
          width={24}
        >
          <span
            className="isvg loading c2 c3"
          />
        </StyledIcon>
      </AlertIcon>
      <StyledBox
        p={2}
        pr={2}
      >
        <div
          className="c4"
        >
          This is a success message
        </div>
      </StyledBox>
    </div>
  </StyledAlert>
</StyledAlert>
`;

exports[`Alert should render a warning alert 1`] = `
.c1 {
  background-color: #fff;
  border: 1px solid #fda509;
  color: #fda509;
  border-radius: 3px;
  font-size: 16px;
  line-height: 1.4;
  max-width: 450px;
  padding: 10px 16px;
  width: 100%;
}

.c1 a {
  color: #fda509;
}

.c4 {
  width: 100%;
  padding: 8px;
  padding-right: 8px;
}

.c3 {
  display: inline-block;
  line-height: 0;
}

.c3 svg {
  height: auto;
  max-height: 100%;
  width: 24px;
}

.c3 svg path {
  fill: #fff;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1;
  padding: 0;
  position: relative;
}

.c2 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  background-color: #fda509;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 16px;
}

<StyledAlert
  outline={true}
  variant="warning"
>
  <StyledAlert
    className="c0"
    outline={true}
    size="md"
    variant="warning"
  >
    <div
      className="c0 c1"
      role="alert"
      size="md"
    >
      <AlertIcon
        color="#fff"
        name="exclamation-circle"
        outline={true}
        variant="warning"
        width={24}
      >
        <StyledIcon
          className="c2"
          color="#fff"
          outline={true}
          src="/media/icons/exclamation-circle.svg"
          variant="warning"
          width={24}
        >
          <span
            className="isvg loading c2 c3"
          />
        </StyledIcon>
      </AlertIcon>
      <StyledBox
        p={2}
        pr={2}
      >
        <div
          className="c4"
        >
          This is a warning message
        </div>
      </StyledBox>
    </div>
  </StyledAlert>
</StyledAlert>
`;

exports[`Alert should render properly 1`] = `
.c1 {
  background-color: #fff;
  border: 1px solid #33b5e5;
  color: #33b5e5;
  border-radius: 3px;
  font-size: 16px;
  line-height: 1.4;
  max-width: 450px;
  padding: 10px 16px;
  width: 100%;
}

.c1 a {
  color: #33b5e5;
}

.c4 {
  width: 100%;
  padding: 8px;
  padding-right: 8px;
}

.c3 {
  display: inline-block;
  line-height: 0;
}

.c3 svg {
  height: auto;
  max-height: 100%;
  width: 24px;
}

.c3 svg path {
  fill: #fff;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1;
  padding: 0;
  position: relative;
}

.c2 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  background-color: #33b5e5;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 16px;
}

<StyledAlert
  outline={true}
  variant="info"
>
  <StyledAlert
    className="c0"
    outline={true}
    size="md"
    variant="info"
  >
    <div
      className="c0 c1"
      role="alert"
      size="md"
    >
      <AlertIcon
        color="#fff"
        name="question-circle"
        outline={true}
        variant="info"
        width={24}
      >
        <StyledIcon
          className="c2"
          color="#fff"
          outline={true}
          src="/media/icons/question-circle.svg"
          variant="info"
          width={24}
        >
          <span
            className="isvg loading c2 c3"
          />
        </StyledIcon>
      </AlertIcon>
      <StyledBox
        p={2}
        pr={2}
      >
        <div
          className="c4"
        >
          Hello World
        </div>
      </StyledBox>
    </div>
  </StyledAlert>
</StyledAlert>
`;
